contracts/infrastructure/dapp/ParaswapFilter.sol

```diff
@@ -33,55 +36,241 @@ interface IParaswap {
+        uniForkInitCode2 = _uniInitCodes[1];
+        uniForkInitCode3 = _uniInitCodes[2];
+
+        for(uint i = 0; i < _adapters.length; i++) {
+            adapters[_adapters[i]] = true;
+            emit AdapterAdded(_adapters[i]); 
+        }
     }
 
-    function isValid(address _wallet, address /*_spender*/, address /*_to*/, bytes calldata _data) external view override returns (bool) {
-        return isMultiSwap(_data) && hasValidBeneficiary(_wallet, _data) && hasTradableToken(_data) && hasValidExchangeAdapters(_data);
+    /**
+     * @notice Add/Remove a DEX adapter to/from the whitelist.
+     * @param _adapters array of DEX adapters to add to (or remove from) the whitelist
+     * @param _authorised array where each entry is true to add the corresponding DEX to the whitelist, false to remove it
+     */
+    function setAuthorised(address[] calldata _adapters, bool[] calldata _authorised) external onlyOwner {
+        for(uint256 i = 0; i < _adapters.length; i++) {
+            if(adapters[_adapters[i]] != _authorised[i]) {
+                adapters[_adapters[i]] = _authorised[i];
+                if(_authorised[i]) { 
+                    emit AdapterAdded(_adapters[i]); 
+                } else { 
+                    emit AdapterRemoved(_adapters[i]);
+                }
+            }
+        }
     }
 
-    function isMultiSwap(bytes calldata _data) internal pure returns (bool) {
+    function isValid(address _wallet, address /*_spender*/, address _to, bytes calldata _data) external view override returns (bool) {
         // disable ETH transfer
         if (_data.length < 4) {
             return false;
         }
-        return getMethod(_data) == MULTISWAP;
+        bytes4 methodId = getMethod(_data);
+        if(methodId == MULTISWAP) {
+            return isValidMultiSwap(_wallet, _data);
+        } 
+        if(methodId == SIMPLESWAP) {
+            return isValidSimpleSwap(_wallet, _to, _data);
+        }
+        if(methodId == SWAP_ON_UNI) {
+            return isValidUniSwap(_to, _data);
+        }
+        if(methodId == SWAP_ON_UNI_FORK) {
+            return isValidUniForkSwap(_to, _data);
+        }
+        if(methodId == MEGASWAP) {
+            return isValidMegaSwap(_wallet, _data);
+        }
+        return false;
     }
 
-    function hasValidBeneficiary(address _wallet, bytes calldata _data) internal pure returns (bool) {
-        (address beneficiary) = abi.decode(_data[228:], (address)); // skipping 4 + 7*32 = 228 bytes
-        return (beneficiary == address(0) || beneficiary == _wallet);
+    function isValidMultiSwap(address _wallet, bytes calldata _data) internal view returns (bool) {
+        (IParaswap.SellData memory sell) = abi.decode(_data[4:], (IParaswap.SellData));
+        return hasValidBeneficiary(_wallet, sell.beneficiary) &&
+            hasTradableToken(sell.path[sell.path.length - 1].to) && 
+            hasValidExchangeAdapters(sell.path);
```

Comment:
Given the comment for `simpleSwap` I believe the `hasValidExchangeAdapters` check is not needed either.